api/protogen/protogen.go

```diff
@@ -267,17 +269,54 @@ func terraformAttributeTypeToProtoType(attrName string, attrType attr.Type) (rep
 	case basetypes.StringType:
 		return false, "string", nil, nil
 	case types.ListType:
-		return terraformRepeatedAttributeTypeToProtoType(attrName, v.ElementType())
+		return terraformRepeatedAttributeTypeToProtoType(resourceName, attrName, v.ElementType(), tagger)
 	case types.SetType:
-		return terraformRepeatedAttributeTypeToProtoType(attrName, v.ElementType())
-	// TODO: Add support for nested objects.
+		return terraformRepeatedAttributeTypeToProtoType(resourceName, attrName, v.ElementType(), tagger)
+	case types.ObjectType:
+		return terraformObjectAttributeTypeToProtoType(resourceName, attrName, v, tagger)
+
 	default:
 		return false, "", nil, fmt.Errorf("unsupported Terraform type: %s", attrType.String())
 	}
 }
 
-func terraformRepeatedAttributeTypeToProtoType(attrName string, elementType attr.Type) (repeated bool, protoType string, messages *message, err error) {
-	elemRepeated, elemProtoType, elemMessage, err := terraformAttributeTypeToProtoType(attrName, elementType)
+// terraformObjectAttributeTypeToProtoType converts a Terraform object attribute into a Protocol Buffer message type.
+func terraformObjectAttributeTypeToProtoType(resourceName, attrName string, obj types.ObjectType, tagger *apiSpecTagger) (repeated bool, protoType string, nestedMessage *message, err error) {
+	messageName := schema.ProtoMessageName(attrName)
+	newMessage := &message{
+		Name:   messageName,
+		Fields: []field{},
+	}
+	wrappedMessageName := resourceName + MessageNameSeperator + messageName
+
+	attrs := schema.SortObjectAttributes(obj.AttrTypes)
+
+	for _, name := range attrs {
+		attrType := obj.AttrTypes[name]
+		isRepeated, t, msg, err := terraformAttributeTypeToProtoType(wrappedMessageName, name, attrType, tagger)
+
+		if err != nil {
+			return false, "", nil, fmt.Errorf("failed to parse field %s in object %s: %w", name, attrName, err)
+		}
+
+		newMessage.Fields = append(newMessage.Fields, field{
+			Repeated: isRepeated,
```

Comment:
Since we can't have explicit Required/Optional flags in the schema for object attributes, we can't do any validation on the provider side for required fields. Therefore, every single object attribute must be considered optional. Make that clear in the generated Protobuf.
```suggestion
			Repeated: isRepeated,
			Optional: true,
```
The server will then have to validate attributes that are actually required.